### PR TITLE
`System::type_id` Consistency

### LINKED ISSUE #11679: Stored `TypeId` of systems have an unexpected value
Ideally we don't need to actually construct the system to get the TypeId here. We should be able to rely on `TypeId::of::<IntoSys::System>()`. However currently that is out of sync with `FunctionSystem`'s `type_id()` implementation. I'll add a `PERF` todo.

_Originally posted by @cart in https://github.com/bevyengine/bevy/pull/8453#discussion_r1476923360_

### PR DESCRIPTION
**Pull request by bushrat011899**
# Objective

- Fixes #11679

## Solution

- Added `IntoSystem::system_type_id` which returns the equivalent of `system.into_system().type_id()` without construction. This allows for getting the `TypeId` of functions (a function is an unnamed type and therefore you cannot call `TypeId::of::<apply_deferred::System>()`)
- Added default implementation of `System::type_id` to ensure consistency between implementations. Some returned `Self`, while others were returning an inner value instead. This ensures consistency with `IntoSystem::system_type_id`.

## Migration Guide

If you use `System::type_id()` on function systems (exclusive or not), ensure you are comparing its value to other `System::type_id()` calls, or `IntoSystem::system_type_id()`.

This code wont require any changes, because `IntoSystem`'s are directly compared to each other.

```rust
fn test_system() {}

let type_id = test_system.type_id();

// ...

// No change required
assert_eq!(test_system.type_id(), type_id);
```

Likewise, this code wont, because `System`'s are directly compared.

```rust
fn test_system() {}

let type_id = IntoSystem::into_system(test_system).type_id();

// ...

// No change required
assert_eq!(IntoSystem::into_system(test_system).type_id(), type_id);
```

The below _does_ require a change, since you're comparing a `System` type to a `IntoSystem` type.

```rust
fn test_system() {}

// Before
assert_eq!(test_system.type_id(), IntoSystem::into_system(test_system).type_id());

// After
assert_eq!(test_system.system_type_id(), IntoSystem::into_system(test_system).type_id());
```

### CODE REVIEW

File: crates/bevy_ecs/src/system/exclusive_function_system.rs [hunk 1]
```diff
@@ -65,11 +65,6 @@ where
         self.system_meta.name.clone()
     }
 
-    #[inline]
-    fn type_id(&self) -> TypeId {
```
Comment (alice-i-cecile) [inline crates/bevy_ecs/src/system/exclusive_function_system.rs:69]: The removal of these methods is breaking. Still good to do, but should be in a migration guide.

**Review by alice-i-cecile — CHANGES_REQUESTED** [review]
We should add a small test to validate this behavior, ideally for all of our different flavors of systems.

File: crates/bevy_ecs/src/system/exclusive_function_system.rs (hunk 1 above)
Comment (bushrat011899) [reply crates/bevy_ecs/src/system/exclusive_function_system.rs:69]: These methods are still available with the same API (they're provided by `System`), they just produce a different value. I'll still add an explicit migration guide.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by bushrat011899** [comment]
I've added some unit tests which cover exclusive function systems, and function systems. They ensure that the `TypeId` value is consistent between `System` and `IntoSystem` implementations, and also consistent with `TypeId::of::<T::System>()` where `T` is the `IntoSystem` type. I've also updated the description of this PR to explicitly call out exactly what cases will require some kind of migration work.

**Review by james7132 — APPROVED** [review]
I'd actually have preferred that we hard-break the existing use cases, since this change is otherwise silent. It's cases like these where I wish we had something similar to Crater to find the affected ecosystem crates. Otherwise though, LGTM.

**Pull request merged**